src/libraries/System.Private.CoreLib/src/System/IO/Strategies/UnixFileStreamStrategy.cs

```diff
@@ -0,0 +1,175 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Threading;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Sources;
+using Microsoft.Win32.SafeHandles;
+
+namespace System.IO.Strategies
+{
+    internal sealed partial class UnixFileStreamStrategy : OSFileStreamStrategy
+    {
+        private ReadAsyncTaskSource? _readAsyncTaskSource;
+
+        internal UnixFileStreamStrategy(SafeFileHandle handle, FileAccess access, FileShare share) :
+            base(handle, access, share)
+        {
+        }
+
+        internal UnixFileStreamStrategy(string path, FileMode mode, FileAccess access, FileShare share, FileOptions options, long preallocationSize) :
+            base(path, mode, access, share, options, preallocationSize)
+        {
+        }
+
+        internal override bool IsAsync => _fileHandle.IsAsync;
+
+        public override IAsyncResult BeginRead(byte[] buffer, int offset, int count, AsyncCallback? callback, object? state) =>
+            TaskToApm.Begin(ReadAsync(buffer, offset, count), callback, state);
+
+        public override int EndRead(IAsyncResult asyncResult) =>
+            TaskToApm.End<int>(asyncResult);
+
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken) =>
+            ReadAsync(new Memory<byte>(buffer, offset, count), cancellationToken).AsTask();
+
+        public override ValueTask<int> ReadAsync(Memory<byte> destination, CancellationToken cancellationToken)
+        {
+            if (!CanRead)
+            {
+                ThrowHelper.ThrowNotSupportedException_UnreadableStream();
+            }
+
+            if (CanSeek)
+            {
+                // This implementation updates the file position after the operation completes, rather than before.
+                // Also, unlike the Net5CompatFileStreamStrategy implementation, this implementation doesn't serialize operations.
```

Comment:
I am a big fan of this change. If we do that for Unix we should most probably do it for Windows as well:

https://github.com/dotnet/runtime/blob/d25af09fba6f10701c62b641227dbe62c649a890/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AsyncWindowsFileStreamStrategy.cs#L38-L55

but not necessarily in this PR